File: ios/PlantPal/PlantPal/Services/EdgeCaseHandler.swift
import Foundation
import SwiftData

final class EdgeCaseHandler {
    
    static func handleDayRollover(plant: Plant, sprite: Sprite, now: Date = .now) {
        let calendar = Calendar.current
        let lastCalcDay = calendar.component(.day, from: plant.lastCalculatedAt)
        let currentDay = calendar.component(.day, from: now)
        
        if lastCalcDay != currentDay {
            resetDailyHabits(plant: plant)
        }
    }
    
    static func resetDailyHabits(plant: Plant) {
        // In real implementation, this would query and reset HabitTask.isCompletedToday
        // The daily reset happens when a new day is detected
    }
    
    static func handleLongAbsence(plant: Plant, sprite: Sprite, now: Date = .now) {
        let elapsedHours = now.timeIntervalSince(plant.lastCalculatedAt) / 3600
        
        if elapsedHours > 72 {
            // After 3+ days absence, cap the damage so plant doesn't instantly die
            // Apply max 48 hours of decay (prevents total death from long absence)
            plant.lastCalculatedAt = now.addingTimeInterval(-min(elapsedHours, 48) * 3600)
        }
    }
    
    static func handleNewPlantCreation(plant: Plant) {
        plant.health = 1.0
        plant.waterLevel = 1.0
        plant.lightLevel = 1.0
        plant.growthProgress = 0.0
        plant.nutrients = 0.0
        plant.plantedAt = Date()
        plant.lastCalculatedAt = Date()
        plant.lastWateredAt = Date()
        plant.lastLightAt = Date()
    }
    
    static func clampAllValues(plant: Plant) {
        plant.health = max(0, min(1, plant.health))
        plant.waterLevel = max(0, min(1, plant.waterLevel))
        plant.lightLevel = max(0, min(1, plant.lightLevel))
        plant.growthProgress = max(0, min(1, plant.growthProgress))
        plant.nutrients = max(0, plant.nutrients)
    }
    
    static func handleSpriteRecovery(sprite: Sprite) {
        if sprite.happiness < 0.1 {
            sprite.happiness = 0.1
        }
    }
}

File: android/app/src/main/java/com/plantpal/engine/EdgeCaseHandler.kt
package com.plantpal.engine

import com.plantpal.data.entity.PlantEntity
import com.plantpal.data.entity.SpriteEntity

object EdgeCaseHandler {

    fun handleLongAbsence(plant: PlantEntity, now: Long = System.currentTimeMillis()): PlantEntity {
        val elapsedHours = (now - plant.lastUpdated) / 3600000.0
        return if (elapsedHours > 72) {
            plant.copy(lastUpdated = now - minOf(elapsedHours, 48.0).toLong() * 3600000)
        } else {
            plant
        }
    }

    fun clampAllValues(plant: PlantEntity): PlantEntity {
        return plant.copy(
            health = plant.health.coerceIn(0.0, 1.0),
            waterLevel = plant.waterLevel.coerceIn(0.0, 1.0),
            lightLevel = plant.lightLevel.coerceIn(0.0, 1.0),
            growthProgress = plant.growthProgress.coerceIn(0.0, 1.0),
            nutrients = maxOf(0.0, plant.nutrients)
        )
    }

    fun handleSpriteRecovery(sprite: SpriteEntity): SpriteEntity {
        return if (sprite.happiness < 0.1) {
            sprite.copy(happiness = 0.1)
        } else {
            sprite
        }
    }

    fun handleNewPlantCreation(): PlantEntity {
        return PlantEntity.create()
    }
}